### ISSUE TITLE
wrong hostname for vsphere rke2 clusternodes in 2.6.4

### ISSUE THREAD

**Issue by Eldritch2004** [rancher/rancher#37412]
In 2.6.3 creating a  rke2 cluster in vSphere used the vm Name as a hostname in cloud-init for Alma Linux 8 and centos8.
The new VM nodes are created from a VM template

So the VM clustername-nodename-f8a6b746-kzwgh would lead to the same hostname and fqdn in the resulting system.

With 2.6.4 I now have the problem that the hostname for the resulting hostname for the new VMs is always the hostname of the VM template and not the name of the VM anymore.

SURE-4636

**Comment by jandunk** [comment]
I'm having the same issue. I suspect it's coming from this rancher-machine commit https://github.com/rancher/machine/commit/d9a83824d326af6b427ceb04388b35c1aa2bb51f by @rosskirkpat(https://github.com/rancher/machine/commits?author=rosskirkpat). The meta-data now contains hostname rather than local-hostname. This gets ignored in my Ubuntu 20.04 template. I have worked around this by making the cloud-config of the node template mount the iso extract the hostname from the meta-data and setting it with hostnamectl.

**Comment by Eldritch2004** [comment]
> I'm having the same issue. I suspect it's coming from this rancher-machine commit [rancher/machine@d9a8382](https://github.com/rancher/machine/commit/d9a83824d326af6b427ceb04388b35c1aa2bb51f) by @rosskirkpat(https://github.com/rancher/machine/commits?author=rosskirkpat). The meta-data now contains hostname rather than local-hostname. This gets ignored in my Ubuntu 20.04 template. I have worked around this by making the cloud-config of the node template mount the iso extract the hostname from the meta-data and setting it with hostnamectl.

Do you have a code example for your workaround?

**Comment by jandunk** [comment]
```
# when rancher includes hostname rather then local-hostname hostname is not set
        sudo mount /dev/sr2 /mnt
        metadata=`sudo cat /mnt/meta-data`
        if [[ $metadata != *"local-hostname"* ]]; then
          echo "Updating hostname because meta-data contains hostname instead of local-hostname"
          HOSTNAME=`sudo cat /mnt/meta-data | grep hostname | awk '{print $2}'`
          sudo hostnamectl set-hostname $HOSTNAME
        fi
```

**Comment by rosskirkpat** [comment]
Thanks for filing this issue @Eldritch2004. When using the NoCloud datasource for metadata, `local-hostname` is the correct parameter and `hostname` is not valid. I will add this to the triage list and ensure it is reviewed in the next sprint. 

One potential workaround is to add one of the following options to your cloud-config depending on if you are using FQDN or short hostnames.

```yml
preserve_hostname: false
prefer_fqdn_over_hostname: true
fqdn: <hostname>
```

```yml
preserve_hostname: false
prefer_fqdn_over_hostname: false
hostname: <hostname>
```

https://github.com/canonical/cloud-init/commit/45db197cfc7e3488baae7dc1053c45da070248f6
> it is distro dependent whether `hostname` or `fqdn` is used,
unless the `prefer_fqdn_over_hostname` option is true and fqdn is set
it will force the use of FQDN in all distros, and if false then it will
force the hostname use.

**Comment by Eldritch2004** [comment]
@rosskirkpat Would love to, but we scripted the whole process and I haven't found a good way to get a unique hostname or fqdn automatically in cloud-init so far, but maybe I just missed something.

**Comment by Jellyfrog** [comment]
Seeing the same,

Rhel 8.5 and 8.6
Rancher 2.6.4

**Comment by Jellyfrog** [comment]
Hackish workaround that seems to work:

When setting up a new cluster, provide an extra cloud-init config for it:
```
#cloud-config

runcmd:
 - 'hostnamectl set-hostname $(cloud-init query -a | grep \"hostname\" | cut -d"\"" -f4)'
 ````

**Comment by fm4991** [comment]
@rosskirkpat any progress on this? maybe without an hackish workaround

**Comment by snasovich** [comment]
rancher/machine to be bumped to `v0.15.0-rancher94` or later once 2.6.7 is released to complete this issue before it can be tested.

**Comment by snasovich** [comment]
@HarrisonWAffel , could you please take care of bumping rancher-machine for this?

**Comment by HarrisonWAffel** [comment]
Machine version within rancher has been bumped to v0.15.0-rancher95, this issue should be OK for testing.

**Comment by HarrisonWAffel** [comment]
### Root Cause

Rancher-machine changed how the hostname was set within the cloud-init when provisioning VSphere nodes. `local-hostname` is [the correct parameter for the NoCloud datasource.](https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html) 

### What was fixed

The change within rancher-machine has been reverted, and now `local-hostname` is once again used instead of `hostname`. 

### Areas to be tested

1. Create a VSphere VM template with RHEL 8.5
2. Create a 3 node cluster using this template 
3. Check that the VM hostname is equal to the name of the node, and not the name of the VM Template

### Potential regressions

Test if similar functionality has been broken for other OS’s, such as Ubuntu or other non RHEL based OS’s (SLES, Leap, etc). 

---

This test-template will be enhanced next week with additional information on the potential regressions and rationale for the initial change.

**Comment by jameson-mcghee** [comment]
When testing on v2.6-head (Commit ID: 68215bc), I was able to reproduce this issue. The Hostname address was ** localhost.localdomain**. After confirming with @HarrisonWAffel, I have reopened this issue.
cc @markusewalker @slickwarren @sowmyav27 

v2.6-head:
![image.png](https://zube.io/files/rancher/e49f76ba85ce88949d44ceace24b4626-image.png)

**Comment by jameson-mcghee** [comment]
For comparison purposes, this was observed on a vSphere Node Driver cluster on v2.6-head (Commit ID: 68215bc).

![image.png](https://zube.io/files/rancher/9384f0a033797d80f38d48ec6d1344e5-image.png)

**Comment by fafarun** [comment]
> [update windows logic rancher/machine#192 ](/rancher/machine/pull/192)
>  Merged
> [<img alt="@snasovich" width="20" height="20" src="https://avatars.githubusercontent.com/u/85187633?s=40&amp;u=2a2f5885af01a7e3829bc92f37caba26f7b9244b&amp;v=4">](/snasovich)

Thanks for the workaround. 
I still have the bug on the 2.6.8 version.

**Comment by HoustonDad** [comment]
We need to get this into 2.6 ASAP, leaving this til 2.7 blocks a LOT of customers / clients.

**Comment by jameson-mcghee** [comment]
**TEST PLAN**

**Intended Test Setup**
-	Rancher version(s): v2.6.8 & v2.7-head
-	Browser type & version: Brave & Most Current
-	Rancher Host: vSphere
-	Rancher Setup (Deployed From Template):
o	Template Name: ubuntu-vm-esxi7-2
o      Operating System: Ubuntu Linux 20.04.4 LTS
o      CPUs: 4
o      Memory: 16 GB
o	Size on Disk: 20 GB
o	Datastore: ranch01-silo02-vm01
-	Deployment Type: Docker
-	User Permission: Standard

**Steps to Reproduce the issue:**
1. In vSphere, create a Rancher v2.6.8 server using the _Intended Test Setup_
2. Also in vSphere, create 3 cluster nodes using the _rhel8_ template
3. Inside Rancher, create a Custom RKE2 cluster, and register the _rhel8_ nodes with 1 etcd, 1 cp, and 1 worker
4. Once the Cluster is Active, navigate to the Cluster Explorer --> Machines page for the cluster, open the YAML view for one of the nodes
  - Note that the address field for the _Hostname_ will state _**localhost.localdomain**_
![image.png](https://zube.io/files/rancher/c8ea725d55e7e46cfaa8efe904204cc0-image.png)

**Comment by HarrisonWAffel** [comment]
After finding time to look more closely at this issue and the related changes I believe that our initial testing was inadequate and did not properly identify the expected behavior after reverting the hostname change. Ross is no longer on our team and some of the particulars of this issue were lost in the transition (we miss you Ross!). Now that I have a better understanding of this issue I'm adding a new, more detailed, test template so that it can be retested. If anyone monitoring this thread has additional information, additional test cases, or corrections to any of the information listed here so far you are encouraged to add them 😃 

### Root cause

For many operating systems using `cloud-init` the hostname can be specified using the `local-hostname` field within the cloud-init meta-data. Others use the `set_hostname` module, which uses the `hostname` field to define a hostname. The correct value to use depends on the version of cloud-init being used. As I understand it, Rancher cannot detect the version of cloud-init being used by a particular machine as it generates the `user-data` and `meta-data` values before powering the virtual machine on. 

Because of this limitation the PR which introduced the regression opted to support the newest versions of cloud-init, which uses the `hostname` field. It was expected that this change would not cause issues with RHEL 8.x, however this thread would suggest otherwise. 


### What was fixed, or what changes have occurred
  The use of `hostname` was reverted and `local-hostname` is now used when creating `meta-data`. 
### Areas or cases that should be tested

@jameson-mcghee I noticed in prior testing that some steps were inaccurate, and this is my fault. This issue and its related fix revolves around node driver provisioned VSphere nodes and as such the VM's must be created by Rancher, not manually. This is likely the reason why some of the test results see `localhost.localdomain` as the hostname. rancher-machine configures the cloud-init `meta-data` before creation, so manually creating the VM and adding it to a custom cluster circumvents rancher-machine entirely. 

+ Create a Rancher serving running 2.6-head or 2.7-head, the particular cloud provider should not matter as long as you can create downstream VSphere clusters
+ configure a VSphere Template within Rancher for the following operating systems 
   + RHEL 7.9
   + RHEL 8.5, RHEL 8.6
   + Alma Linux 8 
   + Centos 8
   + Ubuntu 20.04
   + SLES 15
   + LEAP 15.4
+ Create a cluster for each operating system (1 node, all roles) within Rancher
+ For each cluster, ensure that the hostname is equal to the name of the VM, and not the name of the VSphere template used 
+ Report the clusters which have the appropriate hostname, and which ones do not. 


---

The change for reverting `hostname` to `local-hostname` has already been performed and is in the latest release of rancher-machine, these additional test steps will help us determine if the regression has been fully addressed and if not what extra work is required to regain the functionality present in 2.6.3.

**Comment by jameson-mcghee** [comment]
Per the offline conversation, this issue cannot be further validated without the creation of a valid RHEL 8 vSphere template. Therefore I am moving this ticket to Blocked until the test template(s) needed to validate this ticket can be created.

**Comment by snasovich** [comment]
@fafarun @HoustonDad , is there a chance you could test it on v2.6.9+ ?
There could be some confusion https://github.com/rancher/rancher/pull/38700 showing that it make it into `v2.6.8-rc2` tag but that tag was actually for v2.6.9 (some versioning confusion due to [v2.6.8](https://github.com/rancher/rancher/releases/tag/v2.6.8) being an unplanned emergency release.
The actual bump of `rancher-machine` with the fix (`v0.15.0-rancher95`) is in v2.6.9+ only (see https://github.com/rancher/rancher/blob/v2.6.9/go.mod#L115 vs https://github.com/rancher/rancher/blob/v2.6.8/go.mod#L115).

**Comment by snasovich** [comment]
Closing this issue as we have received confirmation from couple customers that it's working as expected now in 2.6.9+ and 2.7.0+.

@fafarun @HoustonDad , please let us know if you're still running into this issue in later versions.

FYI @daviswill2 @Sahota1225
